apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +80,193 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
 
 	return values
 }
+
+// AddError adds an error-level Subcondition to the DetailedCondition.
+// AddError will also update the DetailedCondition's state to take into account
+// the error that's present.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddError(errorType, reason, message string) {
+
+	// Update the condition so that it indicates there's at least one error
+	// This needs to be here because conditions may be normal-true (positive)
+	// polarity (like `Valid`), or abnormal-true (negative) polarity
+	// (like `ErrorPresent`)
+	if dc.IsPositivePolarity() {
+		dc.Status = ConditionFalse
+	} else {
+		dc.Status = ConditionTrue
+	}
+
+	// Update the DetailedCondition Reason and message if the reason
+	// is different to the current one. We have to namespace
+	// the DetailedCondition reason by the errorType because we're
+	// adding SubConditions; DetailedCondition Reason is a summary of the
+	// SubCondition reasons.
+	detailedReason := errorType + reason
```

Comment:
can we put a `:` or some kind of separator between them?

Reply:
Absolutely, I should have done that.

Reply:
I've added a constant for the Valid type, the only type we are currently using. Changing the type of the DetailedCondition `Type` field is not available to us because we are inlining (what will be) the upstream Condition type.

Reply:
I'm still on the fence about this DetailedCondition API. I think that it's quite specific to the `Valid` condition and I can see how it will work OK there, but I'm not really convinced that this logic belongs in the general `DetailedCondition`.

We are hard up on the release date, so what if we move all this code into the status package as `ValidCondition` type something like this:

```
type ValidCondition interface {
    AddError()
    AddWarning()

    AsDetailedCondition() *projectcontour.DetailedCondition
}
```

Reply:
OK, I thought some more and how about we keep AddWarning/AddError on the DetailedCondition, but we move the rest of the logic to the `status` package, where we can apply the "Valid" conditions transformation based on the known condition types:

```Go
func TransformValidCondition(c *DetailedCondition) {
 assert(c.Type == "Valid")

  switch len(c.Errors) {
  case 0:
    c.Status = ConditionTrue
  case 1:
    c.Status = ConditionFalse
    c.Reason = c.Errors[0].Reason
    c.Message = rollup(c.Errors)
  default:
    c.Status = ConditionFalse
    c.Reason = "MultipleErrors"
    c.Message = rollup(c.Errors)
  }
}
```

🤷

Reply:
Sure, I'll move it all around then. I feel like all that it means is that there will be zero consistency across the use of Subconditions in various places, but whatever.